salt/cache/localfs_key.py

```diff
@@ -0,0 +1,375 @@
+        data = data["pub"]
+    elif bank == "denied_keys":
+        # denied keys is a list post migration, but is a single key in legacy
+        data = data[0]
+        base = "minions_denied"
+    elif bank == "master_keys":
+        # private keys are separate from permissive_pki_access
+        umask = 0o277
+        base = ""
+        # even in clustered mode, master and signing keys live in the
+        # non-clustered pki dir
+        if key in NON_CLUSTERED_MASTER_KEYS:
+            cachedir = __opts__["pki_dir"]
+
+    savefn = Path(cachedir) / base / key
+    base = savefn.parent
+
+    if not clean_path(cachedir, savefn, subdir=True):
+        raise SaltCacheError(f"key {key} is not a valid key path.")
+
+    try:
+        os.makedirs(base)
+    except OSError as exc:
+        if exc.errno != errno.EEXIST:
+            raise SaltCacheError(
+                f"The cache directory, {base}, could not be created: {exc}"
+            )
+
+    # delete current state before re-serializing new state
+    flush(bank, key, cachedir, **kwargs)
+
+    tmpfh, tmpfname = tempfile.mkstemp(dir=base)
+    os.close(tmpfh)
+
+    if user:
+        try:
+            import pwd
+
+            uid = pwd.getpwnam(user).pw_uid
+            os.chown(tmpfname, uid, -1)
+        except (KeyError, ImportError, OSError):
+            # The specified user was not found, allow the backup systems to
+            # report the error
+            pass
+
+    try:
+        with salt.utils.files.set_umask(umask):
+            with salt.utils.files.fopen(tmpfname, "w+b") as fh_:
+                fh_.write(salt.utils.stringutils.to_bytes(data))
+
+            if bank == "master_keys":
+                os.chmod(tmpfname, 0o400)
+
+        # On Windows, os.rename will fail if the destination file exists.
+        salt.utils.atomicfile.atomic_rename(tmpfname, savefn)
+    except OSError as exc:
+        raise SaltCacheError(
+            f"There was an error writing the cache file, base={base}: {exc}"
+        )
+
+
+def fetch(bank, key, cachedir, **kwargs):
+    """
+    Fetch and construct state data for a given minion based on the bank and id
+    """
+    if bank in ["keys", "denied_keys"] and not valid_id(__opts__, key):
+        raise SaltCacheError(f"key {key} is not a valid minion_id")
+
+    if bank not in ["keys", "denied_keys", "master_keys"]:
+        raise SaltCacheError(f"Unrecognized bank: {bank}")
+
+    if not clean_path(cachedir, key, subdir=True):
+        raise SaltCacheError(f"key {key} is not a valid key path.")
+
+    if key == ".key_cache":
+        raise SaltCacheError("trying to read key_cache, there is a bug at call-site")
+    try:
+        if bank == "keys":
+            for state, bank in [
+                ("rejected", "minions_rejected"),
+                ("pending", "minions_pre"),
+                ("accepted", "minions"),
+            ]:
+                keyfile = Path(cachedir, bank, key)
```

Comment:
Needs `clean_path` check like on line [line 180](https://github.com/saltstack/salt/pull/67799/files#diff-a824d74f49f472a38571ecbe483f18286d4b95cd05ac041928b9d475d24cb45eR180)

Reply:
hm, thought that check would cover the below. let me just remove 180 and dupe to the others